L1Trigger/TrackFindingTracklet/src/FPGAWord.cc

```diff
@@ -38,8 +38,8 @@ void FPGAWord::set(int value, int nbits, bool positive, int line, const char* fi
     }
     assert(value < (1 << nbits));
   } else {
-    if (value > (1 << (nbits - 1))) {
-      edm::LogProblem("Tracklet") << "value too large:" << value << " " << (1 << (nbits - 1)) << " (" << file << ":"
+    if (value > (1 << (nbits - 1)) - 1) {
+      edm::LogProblem("Tracklet") << "value too large:" << value << " " << (1 << (nbits - 1)) - 1 << " (" << file << ":"
                                   << line << ")";
     }
     assert(value <= (1 << (nbits - 1)));
```

Comment:
The same fix should be applied to the "assert" statement.

Reply:
I would fix the assert as soon as people had a chance to debug the overflows, elsewise we can't process a single TTbar@200PU event.

Reply:
I confirm Thomas's observation that the z residuals overflow the digitisation range, and  that this was previously not noticed because of the bug fixed by this PR. This may hypothetically lead to disagreement with  the MatchCalculator HLS code, since either the LUT tables containing the digitized matching windows or the test data of matched stubs may be wrong. 

The problem stubs are in layer 3 of the barrel, with the cut at https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTracklet/src/MatchCalculator.cc#L223 being slightly too loose to reject the stubs that are outside the digi range. A solution would be to replace "15.0" by "15.0-0.06" at L685-686 of https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTracklet/interface/Settings.h .  Though ideally a check that the match windows do not exceed the digi range could be added, for example, to TrackletLUT::initmatchcut().

@bryates Since this affects the MatchCalculator, could you please add a fix for this to Thomas's PR?

Reply:
> I confirm Thomas's observation that the z residuals overflow the digitisation range, and that this was previously not noticed because of the bug fixed by this PR. This may hypothetically lead to disagreement with the MatchCalculator HLS code, since either the LUT tables containing the digitized matching windows or the test data of matched stubs may be wrong.
> 
> The problem stubs are in layer 1+2 of the barrel, with the cut at https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTracklet/src/MatchCalculator.cc#L223 being slightly too loose to reject the stubs that are outside the digi range. A solution would be to replace "15.0" by "15.0-0.06" at L685-686 of https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTracklet/interface/Settings.h . Though ideally a check that the match windows do not exceed the digi range could be added, for example, to TrackletLUT::initmatchcut().
> 
> @bryates Since this affects the MatchCalculator, could you please add a fix for this to Thomas's PR?
> 
> The problem projection is in layerdisk=8 of the endcap. I suspect the problem may be that whilst TrackletCalculatorBase.cc checks that the barrel projection izproj lies within the digi range, it doesn't make the same check for the endcap one irprojdisk .
> 
> @aehart Since this may affect agreement with the TC HLS, would you be able to check?

I looked into this and there are three places where we need to fix some <= to be < etc. Made the following changes:

ryd@Anderss-MacBook-Pro test % git diff ../src/TrackletCalculatorBase.cc
diff --git a/L1Trigger/TrackFindingTracklet/src/TrackletCalculatorBase.cc b/L1Trigger/TrackFindingTracklet/src/TrackletCalculatorBase.cc
index 7169ab6f52c..f45f82d7555 100644
--- a/L1Trigger/TrackFindingTracklet/src/TrackletCalculatorBase.cc
+++ b/L1Trigger/TrackFindingTracklet/src/TrackletCalculatorBase.cc
@@ -629,9 +629,10 @@ bool TrackletCalculatorBase::barrelSeeding(const Stub* innerFPGAStub,
         continue;
 
       if (irprojdisk[i] < settings_.rmindisk() / ITC->rD_0_final.K() ||
-          irprojdisk[i] > settings_.rmaxdisk() / ITC->rD_0_final.K())
+          irprojdisk[i] >= settings_.rmaxdisk() / ITC->rD_0_final.K())
         continue;
 
+
       projs[i + N_LAYER].init(settings_,
                               i + N_LAYER,
                               iphiprojdisk[i],
@@ -1022,7 +1023,7 @@ bool TrackletCalculatorBase::diskSeeding(const Stub* innerFPGAStub,
       continue;
 
     //check that r projection in range
-    if (irprojdisk[i] <= 0 || irprojdisk[i] > settings_.rmaxdisk() / ITC->rD_0_final.K())
+    if (irprojdisk[i] <= 0 || irprojdisk[i] >= settings_.rmaxdisk() / ITC->rD_0_final.K())
       continue;
 
     projs[settings_.projdisks(iSeed_, i) + N_LAYER - 1].init(settings_,
@@ -1383,7 +1384,7 @@ bool TrackletCalculatorBase::overlapSeeding(const Stub* innerFPGAStub,
       continue;
 
     //check that r projection in range
-    if (irprojdisk[i] <= 0 || irprojdisk[i] > settings_.rmaxdisk() / ITC->rD_0_final.K())
+    if (irprojdisk[i] <= 0 || irprojdisk[i] >= settings_.rmaxdisk() / ITC->rD_0_final.K())
       continue;
 
     projs[N_LAYER + i + 1].init(settings_,



and


ryd@Anderss-MacBook-Pro test % git diff ../src/MatchCalculator.cc       
diff --git a/L1Trigger/TrackFindingTracklet/src/MatchCalculator.cc b/L1Trigger/TrackFindingTracklet/src/MatchCalculator.cc
index f63e1ece941..41113e143fc 100644
--- a/L1Trigger/TrackFindingTracklet/src/MatchCalculator.cc
+++ b/L1Trigger/TrackFindingTracklet/src/MatchCalculator.cc
@@ -221,7 +221,8 @@ void MatchCalculator::execute(double phioffset) {
       }
 
       bool imatch = (std::abs(ideltaphi) <= (int)phimatchcuttable_.lookup(seedindex)) &&
-                    (std::abs(ideltaz * fact_) <= (int)zmatchcuttable_.lookup(seedindex));
+                    (ideltaz * fact_ < (int)zmatchcuttable_.lookup(seedindex)) &&
+                    (ideltaz * fact_ >= -(int)zmatchcuttable_.lookup(seedindex)) ;
 
       if (settings_.debugTracklet()) {
         edm::LogVerbatim("Tracklet") << getName() << " imatch = " << imatch << " ideltaphi cut " << ideltaphi << " "


and

ryd@Anderss-MacBook-Pro test % git diff ../src/MatchProcessor.cc        
diff --git a/L1Trigger/TrackFindingTracklet/src/MatchProcessor.cc b/L1Trigger/TrackFindingTracklet/src/MatchProcessor.cc
index b4d53b560d0..84d78ca28b5 100644
--- a/L1Trigger/TrackFindingTracklet/src/MatchProcessor.cc
+++ b/L1Trigger/TrackFindingTracklet/src/MatchProcessor.cc
@@ -507,7 +507,8 @@ bool MatchProcessor::matchCalculator(Tracklet* tracklet, const Stub* fpgastub, b
     }
 
     bool imatch = (std::abs(ideltaphi) <= phimatchcuttable_.lookup(seedindex)) &&
-                  (std::abs(ideltaz << dzshift_) <= zmatchcuttable_.lookup(seedindex));
+                  (ideltaz << dzshift_ < zmatchcuttable_.lookup(seedindex)) &&
+                  (ideltaz << dzshift_) >= -zmatchcuttable_.lookup(seedindex));
 
     if (settings_.debugTracklet()) {
       edm::LogVerbatim("Tracklet") << getName() << " imatch = " << imatch << " ideltaphi cut " << ideltaphi << " "


(While looking at this I realized another mistake - the radial position of a projection to the disks is stored as a signed quantity in the code. This should be unsigned as the radius is always positive. I have not looked into fixing this.)

For the changes above I will checkout Thomas branch and apply these fixes. I will see if I can do this today, but I have meetings most afternoon so I might not get to it today.

Reply:
> I confirm Thomas's observation that the z residuals overflow the digitisation range, and that this was previously not noticed because of the bug fixed by this PR. This may hypothetically lead to disagreement with the MatchCalculator HLS code, since either the LUT tables containing the digitized matching windows or the test data of matched stubs may be wrong.
> 
> The problem stubs are in layer 1+2 of the barrel, with the cut at https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTracklet/src/MatchCalculator.cc#L223 being slightly too loose to reject the stubs that are outside the digi range. A solution would be to replace "15.0" by "15.0-0.06" at L685-686 of https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/L1Trigger/TrackFindingTracklet/interface/Settings.h . Though ideally a check that the match windows do not exceed the digi range could be added, for example, to TrackletLUT::initmatchcut().
> 
> @bryates Since this affects the MatchCalculator, could you please add a fix for this to Thomas's PR?
> 
> The problem projection is in layerdisk=8 of the endcap. I suspect the problem may be that whilst TrackletCalculatorBase.cc checks that the barrel projection izproj lies within the digi range, it doesn't make the same check for the endcap one irprojdisk .
> 
> @aehart Since this may affect agreement with the TC HLS, would you be able to check?

I've pushed my changes
https://github.com/cms-L1TK/cmssw/blob/83769fa8cddb52a029ad3e452ed877fb0dbdd6d6/L1Trigger/TrackFindingTracklet/interface/Settings.h#L683-L686